### PR TITLE
Burn confirm

### PR DESCRIPTION
**Pull request by xiaoch05**

### CODE REVIEW

**Comment by xiaoch05** [comment]
@HackFisher 
I want to save the unconfirmed message in the contract. But the dispatch precompile can't return the nonce of the message to contract. A compromise solution is that, use another method in contract invoked by issuing pallet to transfer the nonce into the contract's storage. See https://github.com/darwinia-network/darwinia-bridge-sol/pull/66/files#diff-7adfa07934f1340b6f404f85d52cbbf01f7ee03a7d47db98482abcd41011b6e8R134
what's your idea about this?

File: contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol [hunk 1]
```diff
@@ -9,6 +9,7 @@ import "../interfaces/IERC20.sol";
 contract DarwiniaMappingTokenFactory is Initializable, Ownable {
     address public constant DISPATCH_ENCODER = 0x0000000000000000000000000000000000000018;
     address public constant DISPATCH         = 0x0000000000000000000000000000000000000019;
+    address public constant SYSTEM_ACCOUNT   = 0x6D6F646C6461722f64766D700000000000000000;
```
Comment (HackFisher) [inline contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol:12]: Add comments.

This account is the evm account of some substrate pallet?

File: contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol [hunk 2]
```diff
@@ -85,12 +94,14 @@ contract DarwiniaMappingTokenFactory is Initializable, Ownable {
         tokenToInfo[token] = TokenInfo(eventReceiver, tokenType, backing, source);
 
         (bool encodeSuccess, bytes memory encoded) = DISPATCH_ENCODER.call(
-            abi.encodePacked(eventReceiver, bytes4(keccak256("registered(bytes4,uint32,string,string,uint8,address,address)")),
+            abi.encodePacked(eventReceiver, bytes4(keccak256("token_register_response()")),
```
Comment (HackFisher) [inline contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol:97]: Add comments, who it the eventReceiver?

For encoding messages and confirm to provide message delivery proof.

File: contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol [hunk 3]
```diff
@@ -85,12 +94,14 @@ contract DarwiniaMappingTokenFactory is Initializable, Ownable {
...
                              abi.encode(backing, source, token)));
         require(encodeSuccess, "create: encode dispatch failed");
 
-        (bool success, ) = DISPATCH.call(encoded);
-        require(success, "create: call create erc20 precompile failed");
+        if (encoded.length > 0) {
```
Comment (HackFisher) [inline contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol:101]: Add comment, the name of the dispatch call?

File: contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol [hunk 4]
```diff
@@ -109,16 +120,51 @@ contract DarwiniaMappingTokenFactory is Initializable, Ownable {
         require(info.source != address(0), "token is not created by factory");
         IERC20(token).mint(recipient, amount);
     }
+    
+    // cross transfer to remote chain without waiting any confirm information,
+    // this must request the burn proof can be always verified by the remote chain corrently.
```
Comment (HackFisher) [inline contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol:125]:
```suggestion
    // this require the burn proof can be always verified by the remote chain corrently.
```

File: contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol [hunk 5]
```diff
@@ -109,16 +120,51 @@ contract DarwiniaMappingTokenFactory is Initializable, Ownable {
...
+    // so, here the user's token burned directly.
+    function burnAndRemoteUnlock(uint32 specVersion, uint64 weight, address token, bytes memory recipient, uint256 amount) external payable {
```
Comment (HackFisher) [inline contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol:127]: Do we need to keep two versions of implementation, might be better to remove later?

File: contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol [hunk 6]
```diff
@@ -109,16 +120,51 @@ contract DarwiniaMappingTokenFactory is Initializable, Ownable {
...
+        delete transferUnconfirmed[messageId];
+        emit RemoteUnlockConfirmed(messageId, info.sender, info.token, info.amount, result);
+    }
 
-    function crossTransfer(uint32 specVersion, uint64 weight, address token, bytes memory recipient, uint256 amount) external payable {
+    function burnAndSendProof(uint32 specVersion, uint64 weight, address token, bytes memory recipient, uint256 amount) internal {
         require(amount > 0, "can not transfer amount zero");
         TokenInfo memory info = tokenToInfo[token];
         require(info.source != address(0), "token is not created by factory");
         require(IERC20(token).transferFrom(msg.sender, address(this), amount), "transfer token failed");
```
Comment (HackFisher) [inline contracts/mapping-token/contracts/darwinia/DarwiniaMappingTokenFactory.sol:164]: Add comment,

Lock the fund in this before message on remote backing chain get dispatched successfully and burn finally

**Review by HackFisher — APPROVED** [review]

**Pull request merged**